test/k8s_service_canary_deployment_template_test.go

```diff
@@ -0,0 +1,71 @@
+// +build all tpl
+
+// NOTE: We use build flags to differentiate between template tests and integration tests so that you can conveniently
+// run just the template tests. See the test README for more information.
+
+package test
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/ghodss/yaml"
+	"github.com/gruntwork-io/terratest/modules/helm"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// Test that setting canary.enabled = false will cause the helm template to not render the canary Deployment resource
+func TestK8SServiceCanaryEnabledFalseDoesNotCreateCanaryDeployment(t *testing.T) {
+	t.Parallel()
+
+	helmChartPath, err := filepath.Abs(filepath.Join("..", "charts", "k8s-service"))
+	require.NoError(t, err)
+
+	// We make sure to pass in the linter_values.yaml values file, which we assume has all the required values defined.
+	// We then use SetValues to override all the defaults.
+	options := &helm.Options{
+		ValuesFiles: []string{filepath.Join("..", "charts", "k8s-service", "linter_values.yaml")},
+		SetValues:   map[string]string{"canary.enabled": "false"},
+	}
+	out := helm.RenderTemplate(t, options, helmChartPath, []string{"templates/canarydeployment.yaml"})
+
+	// We take the output and render it to a map to validate it is an empty yaml
+	rendered := map[string]interface{}{}
+	err = yaml.Unmarshal([]byte(out), &rendered)
+	assert.NoError(t, err)
+	assert.Equal(t, len(rendered), 0)
+}
+
+// Test that configuring a canary deployment will render to a manifest with a container that is clearly a canary
+func TestK8SServiceCanaryEnabledCreatesCanaryDeployment(t *testing.T) {
+	t.Parallel()
+
+	helmChartPath, err := filepath.Abs(filepath.Join("..", "charts", "k8s-service"))
+	require.NoError(t, err)
+
+	// We make sure to pass in the linter_values.yaml values file, which we assume has all the required values defined.
+	// We then use SetValues to override all the defaults.
+	options := &helm.Options{
+		ValuesFiles: []string{
+			filepath.Join("..", "charts", "k8s-service", "linter_values.yaml"),
+			filepath.Join("fixtures", "canary_deployment_values.yaml"),
+		},
+	}
+	out := helm.RenderTemplate(t, options, helmChartPath, []string{"templates/canarydeployment.yaml"})
+
+	// We take the output and render it to a map to validate it has created a Horizontal Pod Autoscaler output or not
+	rendered := map[string]interface{}{}
+	err = yaml.Unmarshal([]byte(out), &rendered)
+	assert.NoError(t, err)
+	assert.NotEqual(t, 0, len(rendered))
+
+	// Inspect the name of the rendered canary deployment
+	nameField := rendered["spec"].(map[string]interface{})["template"].(map[string]interface{})["spec"].(map[string]interface{})["containers"].([]interface{})[0].(map[string]interface{})["name"]
+	nameString := fmt.Sprintf("%v", nameField)
+
+	// Ensure the name contains the string "-canary"
+	assert.True(t, strings.Contains(string(nameString), "-canary"))
+}
```

Comment:
If you have time, should at least have one more template test related to checking that the containers for canary come from `canary`, and containers for main deployment come from `containerImage`.

Reply:
I added a new convenience testing function that mirrors loading a deployment struct from the deployment template along with arbitrary set values, but for the canary deployment. Although technically it could be a bit more DRY, I felt it was reasonable and fit the existing pattern of one such convenience function for each template. 

Also added two sanity check tests that both types of deployment end up with the correct label in e890cdb

Reply:
LGTM!

Reply:
Cool - resolving this